### PR TITLE
[Case study] Kapwing

### LINKED ISSUE #6805: [Fugu][Vitals][Runtime] Kapwing - Video Editor - CDS Keynote companion post [12/3/2021]
* Proposed title: Kapwing - Video Editor - CDS Keynote companion post
* Proposed by: odealbuquerque
* Related areas: Fugu/Capabilities,Web Vitals/Performance,JS/WASM
* Content type: Case Study (https://web.dev/handbook/case-study-guidelines/)
* Deadline: 12/3/2021
* Deadline explanation: In line with CDS content plan, the aim is to get this published before the holidays and within a month of CDS Keynote
* [Draft](https://docs.google.com/document/d/1GcluKrLaI9zVFTcq900F6ptzLsIlf6v9GZrL7oA2-kk/edit)
* Content summary: Video editing on the web is more possible than ever with the release of updates to WASM, emscriptem, web codecs. And due to the meteoric rise video consumption with the pandemic and creator platforms like TikTok, YouTube, Netflix, the demand for instant, seamless video creation tools built on a range of devices, right in the browser has never been higher. Video editing use cases has enabled other endless possibilities for building powerful, modern web apps including Audio content like podcasts, Game streaming and beyond - there's so much more that is possible for web developers to build amazing experiences and create value in the world.

### PR DESCRIPTION
**Pull request by heyawhite**
Fixes #6805

Changes proposed in this pull request:

-  Create Kapwing case study based on draft

Preview the case study: https://deploy-preview-6962--web-dev-staging.netlify.app/kapwing/

### CODE REVIEW

**Comment by heyawhite** [comment]
@rachelandrew when you have a moment, I'd appreciate your review. Thanks!

**Comment by mihajlija** [comment]
@heyawhite Rowan will make sure you're added to the email thread to check the details about the title, hero image and the author profile info. I think they were fine with "Kapwing: Powerful video editing for the web". It's an external developer so you'll have to add them to the list, ask for a pic etc.

@rowan-m Pls coordinate this next week and forward me any questions about the process :')

File: src/site/content/en/blog/kapwing/index.md
Comment (rachelandrew) [inline src/site/content/en/blog/kapwing/index.md:2]:
```suggestion
title: 'Kapwing: Powerful video editing for the web'
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:41]:
```suggestion
After that first click, Kapwing users can easily do a whole lot more.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:42]:
```suggestion
They can explore free templates, add new layers of free stock videos, insert
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:52]:
```suggestion
While the web provides unique advantages, it also presents distinct
challenges. Kapwing needs to deliver smooth and precise playback of complex,
multi-layered projects across a wide range of devices and network conditions.
To achieve this, we use a variety of web APIs to achieve our performance and
feature goals.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:90]:
```suggestion
```js
import {DBSchema, openDB, deleteDB, IDBPDatabase} from 'idb';

let openIdb: Promise <IDBPDatabase<Schema>>;

const db =
  (await openDB) <
  Schema >
  (
    'kapwing',
    version, {
      upgrade(db, oldVersion) {
        if (oldVersion >= 1) {
          // assets store schema changed, need to recreate
          db.deleteObjectStore('assets');
        }

        db.createObjectStore('assets', {
          keyPath: 'mediaLibraryID'
        });
      },
      async blocked() {
        await deleteDB('kapwing');
      },
      async blocking() {
        await deleteDB('kapwing');
      },
    }
  );
```
```

Comment (malchata) [reply src/site/content/en/blog/kapwing/index.md:90]: These changes are to adhere with [Google's JavaScript style guide](https://google.github.io/styleguide/jsguide.html).

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:92]:
```suggestion
We pass a version and define an `upgrade` function. This is used for
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:95]: > We also pass error handling callbacks blocked and blocking

This doesn't scan for me, and doesn't seem to relate to anything I can see in the code. This either tells me this detail isn't important from a reader's perspective, or that it could use a bit of shoring up.

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:97]:
```suggestion
found useful in preventing issues for users with unstable systems.

Finally,
note our definition of a primary key `keyPath`. In our case this is a unique
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:106]:
```suggestion
      'assets',
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:116]:
```suggestion
    await (await openIdb).add('assets', idbVideo);
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:118]:
```suggestion
```js
export async function addAsset(mediaLibraryID: string, file: File) {
  return runWithAssetMutex(mediaLibraryID, async () => {
    const assetAlreadyInStore = await (await openIdb).get(
      'assets',
      mediaLibraryID
    );    
    if (assetAlreadyInStore) return;
        
    const idbVideo: IdbVideo = {
      file,
      mediaLibraryID,
    };

    await (await openIdb).add('assets', idbVideo);
  });
}
```
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:121]:
```suggestion
IndexedDB access. This is required for any read-modify-write type operations,
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:154]:
```suggestion
```js
export async function getAsset(
  mediaLibraryID: string,
  source: LayerSource | null | undefined,
  location: string
): Promise<IdbAsset | undefined> {
  let asset: IdbAsset | undefined;
  const { idbCache } = window;
  const assetInCache = idbCache[mediaLibraryID];

  if (assetInCache && assetInCache.status === 'complete') {
    asset = assetInCache.asset;
  } else if (assetInCache && assetInCache.status === 'pending') {
    asset = await new Promise((res) => {
      assetInCache.subscribers.push(res);
    }); 
  } else {
    idbCache[mediaLibraryID] = { subscribers: [], status: 'pending' };
    asset = (await openIdb).get('assets', mediaLibraryID);

    idbCache[mediaLibraryID].asset = asset;
    idbCache[mediaLibraryID].subscribers.forEach((res: any) => {
      res(asset);
    });

    delete (idbCache[mediaLibraryID] as any).subscribers;

    if (asset) {
      idbCache[mediaLibraryID].status = 'complete';
    } else {
      idbCache[mediaLibraryID].status = 'failed';
    }
  } 
  return asset;
} 
```
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:157]:
```suggestion
We have our own data structure, `idbCache`, that is used to minimize IndexedDB
accesses. While IndexedDB is fast, accessing local memory is faster. We
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:161]:
```suggestion
accesses to IndexedDB, which otherwise would be common on load.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:170]: This alt text should be more descriptive. As this stands, a person reliant on assistive technologies may not be able to know what's in the image.

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:175]:
```suggestion
This is a YouTube style video, which is common in our app. The user doesn't
move much throughout the clip, so the timelines visual thumbnails are not
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:236]:
```suggestion
We pass this helper the asset that is stored in IndexedDB. Upon completion we 
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:237]:
```suggestion
will update the asset in IndexedDB as well as our own cache.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:242]:
```suggestion
We gather data about the `audioBuffer` with the `AudioContext` constructor, but 
because we aren't rendering to the device hardware we use the 
`OfflineAudioContext` to render to an `ArrayBuffer` where we will store 
amplitude data.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:261]: Is it supposed to consistently be "WebCodecs" or "web codecs"?

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:310]:
```suggestion
This snippet refers to a `demuxer` class, which we use to encapsulate the 
interface to `MP4Box`. We once again access the asset from IndexedDB. These 
segments are not necessarily stored in byte order, and that the `appendBuffer` 
method returns the offset of the next chunk.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (malchata) [inline src/site/content/en/blog/kapwing/index.md:369]:
```suggestion
  return { type: 'value', value: decodedSample };
```

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [reply src/site/content/en/blog/kapwing/index.md:265]: Good catch, WebCodecs, we want consistency and precision. Glad to have your sharp eye on this.

Would also be good to add links to web.dev posts for anything that we have covered previously. Like this https://web.dev/webcodecs/

Comment (heyawhite) [reply src/site/content/en/blog/kapwing/index.md:265]: Great -- linked out!

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [reply src/site/content/en/blog/kapwing/index.md:99]: You could raise this question with the Kapwing developer in Google Docs, @heyawhite can share. Keep in mind that we are looking to get this out the door as soon as possible and weigh whether it is worth the discussion.

Comment (malchata) [reply src/site/content/en/blog/kapwing/index.md:99]: If we're on a tight timeline, then it's not worth it. It could just be me that's having trouble understanding 😅

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [reply src/site/content/en/blog/kapwing/index.md]: I'm OOO so I don't want to look too closely: if the edits are only changing the formatting to make the code more readable, great! Thank you for the extra attention on this!

If you've suggested any changes that are prescriptive (like use `let` instead of `var`), someone should get Josh from Kapwing to sign off--it is their app and their code after all. @heyawhite @rowan-m to handle the comms around that.

**Comment by mihajlija** [comment]
Looking good! Definitely check about that hero image, we have automatic CDN compression and it doesn't look great atm. 

@heyawhite Quick tip: Use batch commit for suggestions, it keeps the commit history cleaner and the PR more easily scannable https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

I made a note about this in my How we work guide, but didn't get to actually write it down: https://docs.google.com/document/d/1Sgz5_8VxItFtIO_6W2D2Ni9oZxZA1BirPC85aD4UxOQ/edit?resourcekey=0-VJqob8uK0VdCHq5ZTzNXsQ

File: src/site/content/en/blog/kapwing/index.md
Comment (heyawhite) [reply src/site/content/en/blog/kapwing/index.md]: I've updated this significantly:

> Kapwing's editor has a menu for media, including several templates and custom elements, including some templates which are specific to certain platforms like LinkedIn; a timeline which separates video, audio, and animation; canvas editor with export quality options; a preview of the video; and more capabilities.

There's a lot going on, and I don't know that every piece needs to be written out as it's not all useful in the context of this blog post. However, the new copy has significantly more value.

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:18]: @heyawhite 

Add an aside that will introduce Josh, like I did here for Chris https://web.dev/telegraph/.

Here's Josh's LinkedIn https://www.linkedin.com/in/joshua-grossberg-148b7a14

Add that Linkedin link to his author profile as well, so readers could reach out to him directly with questions (or comments).

cc @malchata for another tip for working on case studies

Comment (heyawhite) [reply src/site/content/en/blog/kapwing/index.md:18]: His LinkedIn is in the profile -- not sure why it isn't showing up in the preview. Intro forthcoming

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:432]:
```suggestion
application performance, and to extend functionality with tools like WebGL.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:411]:
```suggestion
1. Creating an `encodedVideoChunk` using data from the demuxer.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:412]:
```suggestion
1. Calling the `decodeEncodedFrame` method.
```

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:286]:
```suggestion
Recently we have moved over to [WebCodecs](/webcodecs/), which
```

File: src/site/content/en/blog/kapwing/index.md
Comment (heyawhite) [reply src/site/content/en/blog/kapwing/index.md:18]: Figured it out -- it was a formatting error (for a few authors, actually). Should be fixed. Got partnerships' LGTM, so now I just need one from @mihajlija or @malchata to publish.

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [reply src/site/content/en/blog/kapwing/index.md:286]: Rule of thumb: use relative links for anything web.dev

**Comment by mihajlija** [comment]
Another link to shim in at the first mention of IndexedDB https://web.dev/storage-for-the-web/

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:440]: @heyawhite  Can you try something to smooth out this abrupt change of tone?

If you have time for it, it's not absolutely necessary, it's just a nice finishing touch and an opportunity for a writing exercise.

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:435]: link to tensorflow pls

File: src/site/content/en/blog/kapwing/index.md
Comment (mihajlija) [inline src/site/content/en/blog/kapwing/index.md:430]: Link to web workers https://web.dev/off-main-thread/

**Review by mihajlija — APPROVED** [review]

**Comment by mihajlija** [comment]
Preemptively approved, pending comments to be resolved before merging. 

*hero image is not great not terrible, but we can ship it 🚢

@heyawhite and @malchata you both did great, I appreciate your help on this!

File: src/site/content/en/blog/kapwing/index.md
Comment (heyawhite) [reply src/site/content/en/blog/kapwing/index.md]: I've clarified this, it is indeed related to the code:

> We pass error-handling callbacks, `blocked` and `blocking`, which we've found useful in preventing issues for users with unstable systems.

Thanks for pointing it out.

**Pull request merged**
